src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs

```diff
@@ -1059,5 +1059,174 @@ public unsafe void CopyFromAscii(string data)
+                        chars,
+                        charIndex,
+                        charLength - charIndex,
+                        true,
+                        out bytesUsed,
+                        out charsUsed,
+                        out completed);
+                    return new string(chars, 0, charIndex + charsUsed);
+                }
+                else if (block.Next == null)
+                {
+                    decoder.Convert(
+                        block.Array,
+                        index,
+                        following,
+                        chars,
+                        charIndex,
+                        charLength - charIndex,
+                        true,
+                        out bytesUsed,
+                        out charsUsed,
+                        out completed);
+                    return new string(chars, 0, charIndex + charsUsed);
+                }
+                else
+                {
+                    decoder.Convert(
+                        block.Array,
+                        index,
+                        following,
+                        chars,
+                        charIndex,
+                        charLength - charIndex,
+                        false,
+                        out bytesUsed,
+                        out charsUsed,
+                        out completed);
+                    charIndex += charsUsed;
+                    remaining -= following;
+                    block = block.Next;
+                    index = block.Start;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ArraySegment<byte> GetArraySegment(MemoryPoolIterator end)
+        {
+            if (IsDefault || end.IsDefault)
+            {
+                return default(ArraySegment<byte>);
+            }
+            if (end.Block == Block)
+            {
+                return new ArraySegment<byte>(Block.Array, Index, end.Index - Index);
+            }
+
+            return GetArraySegmentMultiBlock(ref end);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private ArraySegment<byte> GetArraySegmentMultiBlock(ref MemoryPoolIterator end)
+        {
+            var length = GetLength(end);
+            var array = new byte[length];
+            CopyTo(array, 0, length, out length);
+            return new ArraySegment<byte>(array, 0, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector<byte> GetVector(byte vectorByte)
+        {
+            // Vector<byte> .ctor doesn't become an intrinsic due to detection issue
+            // However this does cause it to become an intrinsic (with additional multiply and reg->reg copy)
+            // https://github.com/dotnet/coreclr/issues/7459#issuecomment-253965670
+            return Vector.AsVectorByte(new Vector<ulong>(vectorByte * 0x0101010101010101ul));
```

Comment:
Couldn't you keep the original static vectors using this? We had a noticeable perf gain back when you changed `Seek()` to take vector refs instead of instantiating new vectors on every call.